### PR TITLE
Make metrics compliant with OpenMetrics' standards

### PR DESCRIPTION
**Pull request by mrueg**
running `promtool check metrics` against the current metric output (2.3.11), it mentions:
```
dovecot_imap_command_count counter metrics should have "_total" suffix
dovecot_imap_command_count non-histogram and non-summary metrics should not have "_count" suffix
dovecot_imap_command_duration_usecs_sum counter metrics should have "_total" suffix
dovecot_imap_command_duration_usecs_sum non-histogram and non-summary metrics should not have "_sum" suffix
dovecot_push_notification_http_finished_count counter metrics should have "_total" suffix
dovecot_push_notification_http_finished_count non-histogram and non-summary metrics should not have "_count" suffix
dovecot_push_notification_http_finished_duration_usecs_sum counter metrics should have "_total" suffix
dovecot_push_notification_http_finished_duration_usecs_sum non-histogram and non-summary metrics should not have "_sum" suffix
dovecot_push_notifications_count counter metrics should have "_total" suffix
dovecot_push_notifications_count non-histogram and non-summary metrics should not have "_count" suffix
dovecot_push_notifications_duration_usecs_sum counter metrics should have "_total" suffix
dovecot_push_notifications_duration_usecs_sum non-histogram and non-summary metrics should not have "_sum" suffix
dovecot_stats_uptime_seconds counter metrics should have "_total" suffix
dovecot_storage_http_gets_count counter metrics should have "_total" suffix
dovecot_storage_http_gets_count non-histogram and non-summary metrics should not have "_count" suffix
dovecot_storage_http_gets_duration_usecs_sum counter metrics should have "_total" suffix
dovecot_storage_http_gets_duration_usecs_sum non-histogram and non-summary metrics should not have "_sum" suffix
```

There was also feedback from a Prometheus' developer in https://github.com/prometheus/docs/pull/1738#issuecomment-683635557

@jeffpc @sirainen

### CODE REVIEW

File: src/stats/stats-service-openmetrics.c [hunk 1]
```diff
@@ -141,8 +141,8 @@ openmetrics_export_metric_value(struct openmetrics_request *req, string_t *out,
 			    timestamp);
 		break;
 	case OPENMETRICS_METRIC_TYPE_DURATION:
-		str_printfa(out, " %"PRIu64" %"PRId64"\n",
-			    stats_dist_get_sum(metric->duration_stats),
+		str_printfa(out, " %f %"PRId64"\n",
+			    stats_dist_get_sum(metric->duration_stats)/1000.0,
```
Comment (cmouse) [inline src/stats/stats-service-openmetrics.c:145]: why is this being divided by 1000?

File: src/stats/stats-service-openmetrics.c [hunk 2]
```diff
@@ -210,7 +210,7 @@ openmetrics_export_histogram(struct openmetrics_request *req, string_t *out,
 			openmetrics_find_histogram_bucket(metric, i);
 
 		if (sub_metric != NULL) {
-			sum += stats_dist_get_sum(sub_metric->duration_stats);
+			sum += stats_dist_get_sum(sub_metric->duration_stats)/1000.0;
```
Comment (cmouse) [inline src/stats/stats-service-openmetrics.c:213]: why is this being divided by 1000? also these divisions should be in separate commit.

**Review by cmouse — CHANGES_REQUESTED** [review]

File: src/stats/stats-service-openmetrics.c [hunk 3]
```diff
@@ -141,8 +141,8 @@ openmetrics_export_metric_value(struct openmetrics_request *req, string_t *out,
 			    timestamp);
 		break;
 	case OPENMETRICS_METRIC_TYPE_DURATION:
-		str_printfa(out, " %"PRIu64" %"PRId64"\n",
-			    stats_dist_get_sum(metric->duration_stats),
+		str_printfa(out, " %f %"PRId64"\n",
```
Comment (sirainen) [inline src/stats/stats-service-openmetrics.c:144]: Maybe %.3f?

(Otherwise I'll leave this to others, since I haven't really looked at openmetrics myself.)

File: src/stats/stats-service-openmetrics.c (hunk 1 above)
Comment (mrueg) [reply src/stats/stats-service-openmetrics.c:145]: To convert from milliseconds to seconds, which is the recommended unit for prometheus metrics. https://prometheus.io/docs/practices/naming/#base-units

File: src/stats/stats-service-openmetrics.c (hunk 3 above)
Comment (mrueg) [reply src/stats/stats-service-openmetrics.c:144]: Thanks, makes sense, since we only have ms precision.

File: src/stats/stats-service-openmetrics.c (hunk 2 above)
Comment (mrueg) [reply src/stats/stats-service-openmetrics.c:213]: I can add a comment to it, why it's converted.

File: src/stats/stats-service-openmetrics.c [hunk 4]
```diff
@@ -14,7 +14,7 @@
 #include "stats-metrics.h"
 #include "stats-service-private.h"
 
-#define OPENMETRICS_CONTENT_VERSION "0.0.4"
+#define OPENMETRICS_CONTENT_VERSION "0.0.5"
```
Comment (brian-brazil) [inline src/stats/stats-service-openmetrics.c:17]: There is no Openmetrics 0.0.5.

**Review by brian-brazil — COMMENTED** [review]
This is still not valid OpenMetrics format, note that the Prometheus text format and OpenMetrics are different things.

File: src/stats/stats-service-openmetrics.c (hunk 4 above)
Comment (mrueg) [reply src/stats/stats-service-openmetrics.c]: Thanks, fixed!

File: src/stats/stats-service-openmetrics.c (hunk 3 above)
Comment (mrueg) [reply src/stats/stats-service-openmetrics.c:145]: Applied a change to .6f

**Comment by mrueg** [comment]
This is how metrics currently look like, the suggested test in https://www.robustperception.io/checking-openmetrics-output-is-valid passes. The only item I'm not super sure about here are the timestamps. Do they need to be divided by 10e6 (I think the suggestion is to have them in seconds as well)?
```
# HELP process_start_time_seconds Dovecot stats service uptime
# TYPE process_start_time_seconds gauge
process_start_time_seconds 778 1599262064097
# HELP dovecot_build Dovecot build information
# TYPE dovecot_build info
dovecot_build_info{version="2.3.11.3",revision="502c39af9"} 1 1599262064097
# HELP dovecot_storage_http_gets Total number of all events of this kind
# TYPE dovecot_storage_http_gets counter
dovecot_storage_http_gets_total{method="get"} 0 1599262064097
# HELP dovecot_storage_http_gets_duration_seconds Total duration of all events of this kind
# TYPE dovecot_storage_http_gets_duration_seconds counter
dovecot_storage_http_gets_duration_seconds_total{method="get"} 0.000000 1599262064097
# HELP dovecot_imap_command Total number of all events of this kind
# TYPE dovecot_imap_command counter
dovecot_imap_command_total 0 1599262064097
# HELP dovecot_imap_command_duration_seconds Total duration of all events of this kind
# TYPE dovecot_imap_command_duration_seconds counter
dovecot_imap_command_duration_seconds_total 0.000000 1599262064097
# HELP dovecot_push_notifications Total number of all events of this kind
# TYPE dovecot_push_notifications counter
dovecot_push_notifications_total 0 1599262064097
# HELP dovecot_push_notifications_duration_seconds Total duration of all events of this kind
# TYPE dovecot_push_notifications_duration_seconds counter
dovecot_push_notifications_duration_seconds_total 0.000000 1599262064097
# HELP dovecot_push_notification_http_finished Total number of all events of this kind
# TYPE dovecot_push_notification_http_finished counter
dovecot_push_notification_http_finished_total 0 1599262064097
# HELP dovecot_push_notification_http_finished_duration_seconds Total duration of all events of this kind
# TYPE dovecot_push_notification_http_finished_duration_seconds counter
dovecot_push_notification_http_finished_duration_seconds_total 0.000000 1599262064097
# EOF
```

**Comment by brian-brazil** [comment]
> The only item I'm not super sure about here are the timestamps. Do they need to be divided by 10e6 (I think the suggestion is to have them in seconds as well)?

The timestamps should be removed. The content type is also not right for openmetrics.

**Comment by mrueg** [comment]
Maybe I misunderstood the content type from the client_python implemention https://github.com/prometheus/client_python/blob/3cb4c9247f3f08dfbe650b6bdf1f53aa5f6683c1/prometheus_client/openmetrics/exposition.py#L7 I reverted it back to text/plain and dropped the timestamps:

```
# HELP process_start_time_seconds Dovecot stats service uptime
# TYPE process_start_time_seconds gauge
process_start_time_seconds 836
# HELP dovecot_build Dovecot build information
# TYPE dovecot_build info
dovecot_build_info{version="2.3.11.3",revision="502c39af9"} 1
# HELP dovecot_storage_http_gets Total number of all events of this kind
# TYPE dovecot_storage_http_gets counter
dovecot_storage_http_gets_total 0
# HELP dovecot_storage_http_gets_duration_seconds Total duration of all events of this kind
# TYPE dovecot_storage_http_gets_duration_seconds counter
dovecot_storage_http_gets_duration_seconds_total 0.000000
# HELP dovecot_imap_command Total number of all events of this kind
# TYPE dovecot_imap_command counter
dovecot_imap_command_total{cmd_name="EXAMINE"} 3
dovecot_imap_command_total{cmd_name="EXAMINE",tagged_reply_state="OK"} 3
dovecot_imap_command_total{cmd_name="UID SEARCH"} 3
dovecot_imap_command_total{cmd_name="UID SEARCH",tagged_reply_state="OK"} 3
dovecot_imap_command_total{cmd_name="UID FETCH"} 13
dovecot_imap_command_total{cmd_name="UID FETCH",tagged_reply_state="OK"} 13
dovecot_imap_command_total{cmd_name="NOOP"} 20
dovecot_imap_command_total{cmd_name="NOOP",tagged_reply_state="OK"} 20
dovecot_imap_command_total{cmd_name="NAMESPACE"} 1
dovecot_imap_command_total{cmd_name="NAMESPACE",tagged_reply_state="OK"} 1
dovecot_imap_command_total{cmd_name="SEARCH"} 1
dovecot_imap_command_total{cmd_name="SEARCH",tagged_reply_state="OK"} 1
dovecot_imap_command_total{cmd_name="SELECT"} 8
dovecot_imap_command_total{cmd_name="SELECT",tagged_reply_state="OK"} 8
dovecot_imap_command_total{cmd_name="UID STORE"} 6
dovecot_imap_command_total{cmd_name="UID STORE",tagged_reply_state="OK"} 6
dovecot_imap_command_total{cmd_name="STATUS"} 3
dovecot_imap_command_total{cmd_name="STATUS",tagged_reply_state="OK"} 3
dovecot_imap_command_total{cmd_name="UID COPY"} 2
dovecot_imap_command_total{cmd_name="UID COPY",tagged_reply_state="OK"} 2
dovecot_imap_command_total{cmd_name="UID EXPUNGE"} 2
dovecot_imap_command_total{cmd_name="UID EXPUNGE",tagged_reply_state="OK"} 2
# HELP dovecot_imap_command_duration_seconds Total duration of all events of this kind
# TYPE dovecot_imap_command_duration_seconds counter
dovecot_imap_command_duration_seconds_total{cmd_name="EXAMINE"} 0.153950
dovecot_imap_command_duration_seconds_total{cmd_name="EXAMINE",tagged_reply_state="OK"} 0.153950
dovecot_imap_command_duration_seconds_total{cmd_name="UID SEARCH"} 0.024290
dovecot_imap_command_duration_seconds_total{cmd_name="UID SEARCH",tagged_reply_state="OK"} 0.024290
dovecot_imap_command_duration_seconds_total{cmd_name="UID FETCH"} 0.973220
dovecot_imap_command_duration_seconds_total{cmd_name="UID FETCH",tagged_reply_state="OK"} 0.973220
dovecot_imap_command_duration_seconds_total{cmd_name="NOOP"} 0.219900
dovecot_imap_command_duration_seconds_total{cmd_name="NOOP",tagged_reply_state="OK"} 0.219900
dovecot_imap_command_duration_seconds_total{cmd_name="NAMESPACE"} 0.001320
dovecot_imap_command_duration_seconds_total{cmd_name="NAMESPACE",tagged_reply_state="OK"} 0.001320
dovecot_imap_command_duration_seconds_total{cmd_name="SEARCH"} 0.032590
dovecot_imap_command_duration_seconds_total{cmd_name="SEARCH",tagged_reply_state="OK"} 0.032590
dovecot_imap_command_duration_seconds_total{cmd_name="SELECT"} 8.019990
dovecot_imap_command_duration_seconds_total{cmd_name="SELECT",tagged_reply_state="OK"} 8.019990
dovecot_imap_command_duration_seconds_total{cmd_name="UID STORE"} 0.188720
dovecot_imap_command_duration_seconds_total{cmd_name="UID STORE",tagged_reply_state="OK"} 0.188720
dovecot_imap_command_duration_seconds_total{cmd_name="STATUS"} 1.703130
dovecot_imap_command_duration_seconds_total{cmd_name="STATUS",tagged_reply_state="OK"} 1.703130
dovecot_imap_command_duration_seconds_total{cmd_name="UID COPY"} 1.537030
dovecot_imap_command_duration_seconds_total{cmd_name="UID COPY",tagged_reply_state="OK"} 1.537030
dovecot_imap_command_duration_seconds_total{cmd_name="UID EXPUNGE"} 0.054430
dovecot_imap_command_duration_seconds_total{cmd_name="UID EXPUNGE",tagged_reply_state="OK"} 0.054430
# HELP dovecot_push_notifications Total number of all events of this kind
# TYPE dovecot_push_notifications counter
dovecot_push_notifications_total 0
# HELP dovecot_push_notifications_duration_seconds Total duration of all events of this kind
# TYPE dovecot_push_notifications_duration_seconds counter
dovecot_push_notifications_duration_seconds_total 0.000000
# HELP dovecot_push_notification_http_finished Total number of all events of this kind
# TYPE dovecot_push_notification_http_finished counter
dovecot_push_notification_http_finished_total 0
# HELP dovecot_push_notification_http_finished_duration_seconds Total duration of all events of this kind
# TYPE dovecot_push_notification_http_finished_duration_seconds counter
dovecot_push_notification_http_finished_duration_seconds_total 0.000000
# EOF
```

**Comment by brian-brazil** [comment]
> Maybe I misunderstood the content type from the client_python implemention

That's the correct content type for OpenMetrics, so you should be using it.

> process_start_time_seconds 836

This should be a timestamp.

> dovecot_imap_command_total{cmd_name="EXAMINE"} 3
> dovecot_imap_command_total{cmd_name="EXAMINE",tagged_reply_state="OK"} 3

These still have inconsistent labels.

> dovecot_push_notifications and dovecot_push_notifications_duration_seconds_total

These should probably be a summary, as they appear linked.

**Comment by mrueg** [comment]
> > Maybe I misunderstood the content type from the client_python implemention
> 
> That's the correct content type for OpenMetrics, so you should be using it.
> 
So just to get it right this time, re-reverting https://github.com/dovecot/core/pull/136/commits/904fd8bb6bbc485430b5178713b26ba4a4c1601d is correct? 
> > process_start_time_seconds 836
> 
> This should be a timestamp.
> 
Thanks, will fix!
> > dovecot_imap_command_total{cmd_name="EXAMINE"} 3
> > dovecot_imap_command_total{cmd_name="EXAMINE",tagged_reply_state="OK"} 3
> 
> These still have inconsistent labels.
> 
I see.  Unfortunately, this is based on the metrics configuration, there's a group_by option that can generate sub metrics.
If I had failed requests,  `dovecot_imap_command_total{cmd_name="EXAMINE",tagged_reply_state="FAILED"} 3` would appear.

> > dovecot_push_notifications and dovecot_push_notifications_duration_seconds_total
> 
> These should probably be a summary, as they appear linked.
Currently this library only implements histograms and counters. It's a bit out of (my) scope to implement further items.

**Comment by brian-brazil** [comment]
> So just to get it right this time, re-reverting 904fd8b is correct?

Yes, the original was right for OpenMetrics.

> I see. Unfortunately, this is based on the metrics configuration, there's a group_by option that can generate sub metrics.
If I had failed requests, dovecot_imap_command_total{cmd_name="EXAMINE",tagged_reply_state="FAILED"} 3 would appear.

As-is these metrics will be very tricky to use correctly, due to both the double counting and the time series not always being present. This is in addition to breaking latency out by failure/success often resulting in misleading graphs. I'd recommend exporting it only without the tagged_reply_state label.

**Comment by jeffpc** [comment]
> > I see. Unfortunately, this is based on the metrics configuration, there's a group_by option that can generate sub metrics.
> > If I had failed requests, dovecot_imap_command_total{cmd_name="EXAMINE",tagged_reply_state="FAILED"} 3 would appear.
> 
> As-is these metrics will be very tricky to use correctly, due to both the double counting and the time series not always being present. This is in addition to breaking latency out by failure/success often resulting in misleading graphs. I'd recommend exporting it only without the tagged_reply_state label.

Hm, the whole point of the group_by config option (which list a sysadmin-supplied list of event fields to group by) is that sysadmins want stats broken down by the various fields.  For example, in the example above, (I'm assuming) the sysadmin wants to see stats per command & whether or not the command succeeded.

What do you mean by "time series not always being present"? Are you talking about some of the stats lines having only cmd_name while others have cmd_name *and* tagged_reply_state?  Would only including the lines with *all* the fields address your concern? (It'd mean that if one wants to see stats for a particular command - e.g., EXAMINE - one would have to sum up the stats for that command - EXAMINE + OK, EXAMINE + FAILED, etc.)

Is the double counting you're referring to the same problem? Namely that (currently) the output includes "intermediate levels" instead of just the most specific set of labels?

**Comment by brian-brazil** [comment]
> What do you mean by "time series not always being present"? 

There's only a series being exposed for the OK state, but not any other state. In principle all states should always be exposed.

> Would only including the lines with all the fields address your concern?

It would resolve one issue.

> Is the double counting you're referring to the same problem? Namely that (currently) the output includes "intermediate levels" instead of just the most specific set of labels?

The double counting is the intermediate levels.

**Comment by jeffpc** [comment]
> > What do you mean by "time series not always being present"?
> 
> There's only a series being exposed for the OK state, but not any other state. In principle all states should always be exposed.

Ah, I didn't get that before. Hm, this is very hard since the whole point of the group_by syntax and functionality is that the states do not have to be known apriori - saving memory, etc. For example, it'd be silly to keep stats for each user account when we can only keep track of active accounts.

Is there some "proper" way these not-fully-known states spaces are supposed to be exported?

> > Would only including the lines with all the fields address your concern?
> 
> It would resolve one issue.
> 
> > Is the double counting you're referring to the same problem? Namely that (currently) the output includes "intermediate levels" instead of just the most specific set of labels?
> 
> The double counting is the intermediate levels.

Ok.  That's easy enough to fix - just exporting the submetrics that do not have submetrics of their own.

Thanks.

**Comment by jeffpc** [comment]
Internal ticket id: DOP-2102

**Comment by brian-brazil** [comment]
> Is there some "proper" way these not-fully-known state spaces are supposed to be exported?

There's not, I'd expect something like error states to be knowable though. Per user account metrics would be unwise.

**Comment by jeffpc** [comment]
> > Is there some "proper" way these not-fully-known state spaces are supposed to be exported?
> 
> There's not, I'd expect something like error states to be knowable though. Per user account metrics would be unwise.

In theory, yes, errors are knowable.  The group-by functionality aggregates based on event fields which can hold anything - errors, user accounts, IP addresses, etc. All (mostly) represented by strings or integers.  Each deployment cares about slightly different things, and instead of making an ever increasingly complex stats tracking, we made it completely generic.  It is up to the sysadmin to select which field should be used for stats.  Unfortunately, this decoupling of the value from the meaning means that we have no idea about the state space we're dealing with.

For example, here's the list of fields for the imap_command_finished event:

https://doc.dovecot.org/admin_manual/list_of_events/#imap-command-finished

Anyway, thanks for the info.  I think we'll just have to deviate from this part of the openmetrics spec.

**Comment by mrueg** [comment]
@jeffpc anything that's left to be fixed up on this PR? I'm running this version on my dovecot instance and it seems to behave fine.

**Comment by jeffpc** [comment]
We created an internal merge request with this. We'll finish it up and do some cleanups. Thanks.

**Comment by mrueg** [comment]
@jeffpc do you have a rough estimate which release the internal MR will target?

**Comment by sirainen** [comment]
The target is v2.3.14 release. We'll release v2.3.13 soon (next week?) The schedule for v2.3.14 is somewhat unknown.

**Comment by cmouse** [comment]
This has been merged externally.
